### DISCUSSION TITLE
Packaging Seq app with Win-x64 and Linux-x64 runtimes

### DISCUSSION THREAD

**Discussion by MattMofDoom** [datalust/seq-tickets#1356] [General]
Hiya,

I've been bashing my head against this for a while. With my recent contributions to https://github.com/Hinni/seq-input-mssql, @hinni and I were trying to get this running under both Windows and Docker. We came to a roadblock when it comes to integrated authentication, so we published the updates reflecting that it was Windows-only.

Typical error was System.Data.SqlClient.SqlException (0x80131904): Unable to load DLL 'System.Net.Security.Native' or one of its dependencies under Docker or Windows, depending on what we targeted the RID at. I also bumped up against System.PlatformNotSupportedException: Strings.PlatformNotSupported_DataSqlClient in some scenarios.

The thing is that the resulting package does actually appear work under Docker - it's only when trusted connection (integrated auth) that seems to really give us headaches. I realise that this is because when we compile with win-x64, we're getting a default System.Security.Principal.Windows that derives from .NET Framework, while compiling with linux-x64 or both drops this. 

I think that if we could pack the Microsoft.Data.SqlClient runtime-specific binaries for Windows and Linux, this could work on both platforms. That doesn't solve the need for Kerberos configuration etc under Linux for it to work, but it would make things consistent.

As Seq requires us to package the dependencies to be able to run the application, this has posed some complexity that I haven't yet solved. The closest I've come is with https://github.com/novotnyllc/MSBuildSdkExtras where I can at least produce runtime/win-x64 and runtime/linux-x64 packages in the package ... but that's not sufficient for Seq.

Looking for ideas on how to approach this in a way that works for a Seq app. All the various iterations of packaging properties have drained my brain 😂

Cheers,

Matt

**Comment by nblumhardt** [comment]
Thanks for the notes! Yes, this is a bit of a problem for apps built for Seq using .NET, which we're yet to solve.

The underlying app hosting mechanism in Seq does support platform-specific binaries, but only for executables:

https://github.com/datalust/seq-app-jsonarchive/blob/dev/jsonarchive.exe.d.json#L7

Unfortunately, doing this for a .NET app would mean packaging the entire set of .NET runtime dependencies into each platform sub-directory (self-contained publishing), which would bloat out the package quite a bit. I've had mixed success with trimming in the past, but you never know - this might actually become viable at some point... 🤔 

Producing platform-specific executables for Seq Apps does seem to fit a little better with the publishing options provided by the .NET SDK, so it's one avenue we'd be up for exploring, but while we're in the final stages of the 2021.3 release we're a bit short of time :-)

**Reply by MattMofDoom** [reply]
Understood re 2021.3 😁

I figured I'd just pose an open question to see if there were any ideas or pre-existing knowledge in the community that I could try out!

Thanks for the pointer - I'd poked around in a few packages to see if anyone else had solved it, but I didn't think of JsonArchive ... I'll take a look and see if it's possible. Seq.Input.MSSQL is already a bit bloated compared to other apps, since it needs a bunch of dependencies, but the capabilities it provides might be worth a bit of bloat.

**Reply by MattMofDoom** [reply]
Hmm, not sure I can make this work as things stand. I might be missing something, but will park it for now. 

I don't really think there's a ready answer to making integrated authentication working on Docker even if packaging could work. As far as I can see, this requires Kerberos to be set up on the target-  not completely out of the question, but not exactly something we can guarantee will work 'out of the box' with the app installed. 

Might be worth circling around on the authentication story for Azure AD to make it possible to operate against Azure SQL databases, but I don't think there's any provision for this to work as an alternative for on premise integrated authentication.
